Userland/Libraries/LibGPU/ImageFormat.h

```diff
@@ -1,43 +1,179 @@
+enum class PixelComponentBits {
+    AllBits,
+    B1_5_5_5,
+    B2_3_3,
+    B2_10_10_10,
+    B3_3_2,
+    B4_4_4_4,
+    B5_5_5_1,
+    B5_6_5,
+    B8_8_8_8,
+    B10_10_10_2,
+};
+
+// The base data type used as pixel storage
+enum class PixelDataType {
+    Bitmap,
+    Byte,
+    Float,
+    HalfFloat,
+    Int,
+    Short,
+    UnsignedByte,
+    UnsignedInt,
+    UnsignedShort,
+};
+
+// Order of components within a single pixel
+enum class ComponentsOrder {
+    Normal,
+    Reversed,
+};
+
+struct PixelType final {
+    PixelFormat format;
+    PixelComponentBits bits;
+    PixelDataType data_type;
+    ComponentsOrder components_order { ComponentsOrder::Normal };
+};
+
+static constexpr int number_of_components(PixelFormat format)
 {
     switch (format) {
-    case ImageFormat::L8:
+    case PixelFormat::Alpha:
+    case PixelFormat::Blue:
+    case PixelFormat::ColorIndex:
+    case PixelFormat::DepthComponent:
+    case PixelFormat::Green:
+    case PixelFormat::Luminance:
+    case PixelFormat::Red:
+    case PixelFormat::StencilIndex:
         return 1;
-    case ImageFormat::RGB565:
-    case ImageFormat::L8A8:
+    case PixelFormat::LuminanceAlpha:
         return 2;
-    case ImageFormat::RGB888:
-    case ImageFormat::BGR888:
+    case PixelFormat::BGR:
+    case PixelFormat::RGB:
+        return 3;
+    case PixelFormat::BGRA:
+    case PixelFormat::RGBA:
+        return 4;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+static constexpr int number_of_components(PixelComponentBits bits)
+{
+    switch (bits) {
+    case PixelComponentBits::AllBits:
+        return 1;
+    case PixelComponentBits::B2_3_3:
+    case PixelComponentBits::B3_3_2:
+    case PixelComponentBits::B5_6_5:
         return 3;
-    case ImageFormat::RGBA8888:
-    case ImageFormat::BGRA8888:
+    case PixelComponentBits::B1_5_5_5:
+    case PixelComponentBits::B2_10_10_10:
+    case PixelComponentBits::B4_4_4_4:
+    case PixelComponentBits::B5_5_5_1:
+    case PixelComponentBits::B8_8_8_8:
+    case PixelComponentBits::B10_10_10_2:
         return 4;
     default:
         VERIFY_NOT_REACHED();
     }
 }
 
+static constexpr Array<u8, 4> pixel_component_bitfield_lengths(PixelComponentBits bits)
+{
+    VERIFY(bits != PixelComponentBits::AllBits);
+    switch (bits) {
+    case PixelComponentBits::B1_5_5_5:
+        return { 1, 5, 5, 5 };
+    case PixelComponentBits::B2_3_3:
+        return { 2, 3, 3 };
+    case PixelComponentBits::B2_10_10_10:
+        return { 2, 10, 10, 10 };
+    case PixelComponentBits::B3_3_2:
+        return { 3, 3, 2 };
+    case PixelComponentBits::B4_4_4_4:
+        return { 4, 4, 4, 4 };
+    case PixelComponentBits::B5_5_5_1:
+        return { 5, 5, 5, 1 };
+    case PixelComponentBits::B5_6_5:
+        return { 5, 6, 5 };
+    case PixelComponentBits::B8_8_8_8:
+        return { 8, 8, 8, 8 };
+    case PixelComponentBits::B10_10_10_2:
+        return { 10, 10, 10, 2 };
+    default:
```

Comment:
Note to self: remove all `default:` cases from enum switches to make use of compiler errors in the future